### PR TITLE
fix: normalize packed broadcast tensors onto CUDA before NCCL broadcast

### PR DESCRIPTION
**Pull request by mohamed-alansary**
## Summary

This PR makes producer-side packed broadcast robust to mixed CPU/CUDA tensor pieces by normalizing packing tensors onto the current CUDA device before `torch.cat` and NCCL broadcast.

## Motivation

Packed broadcast can receive tensors from mixed devices during weight sync/refit flows. `torch.cat` requires all pieces to be on one device, and NCCL broadcast requires CUDA tensors. Normalizing before packing avoids failures in that path while preserving CPU-only behavior when CUDA is unavailable.

## Changes

- Add a private helper to normalize packed tensor pieces onto the current CUDA device.
- Apply the helper before both producer-side packing flushes.
- Keep CPU-only behavior unchanged when CUDA is unavailable.
- Add regression coverage for:
  - CPU-only helper behavior.
  - CPU-to-CUDA normalization.
  - Mixed CPU/CUDA normalization.
  - Producer mixed-device final flush.
  - Producer mixed-device normal flush.

## Tests

- `python -m pytest tests/unit/utils/test_packed_tensor.py`
  - `9 passed`
- `ruff check nemo_rl/utils/packed_tensor.py tests/unit/utils/test_packed_tensor.py`
  - `All checks passed!`

### CODE REVIEW

**Review by yuki-97 — COMMENTED** [review]
@youngeunkwon0405 could you help review?

@mohamed-alansary just curious, where did you meet this case that need mixed type of tensor?

**Comment by mohamed-alansary** [comment]
Thanks @yuki-97 . To clarify, this is a mixed-device case rather than mixed dtype.

I hit this in a downstream weight sync/refit flow using packed broadcast, where the producer-side iterator/post_iter_func can return some tensors on CPU and others on CUDA. The producer then converts them to flattened uint8 views and calls torch.cat before NCCL broadcast; torch.cat requires all pieces to be on one device, so mixed CPU/CUDA pieces fail before the broadcast.

This PR normalizes the packed pieces onto a CUDA device when CUDA is available, while leaving the no-CUDA path unchanged. I also added regression coverage for both the final-flush and normal-flush producer paths.

**Comment by youngeunkwon0405** [comment]
Hi @mohamed-alansary, I am also curious when it could be the case that the producer iterator returns a CPU tensor? Can you share the case in more detail?

I think if there is the case, the right implemenation should be
Changing from
```
tensor = (
                        post_iter_func(next(iterator))
                        .contiguous()
                        .view(torch.uint8)
                        .view(-1)
                    )
```
to
```
tensor = (
                        post_iter_func(next(iterator))
                        .contiguous()
                        .to(torch.device("cuda"))
                        .view(torch.uint8)
                        .view(-1)
                    )
```

**Comment by mohamed-alansary** [comment]
Thanks @youngeunkwon0405 . The case I hit was during non-colocated server-generation refit / weight sync. In that path the producer streams state-dict / adapter tensors into packed broadcast, and some pieces were materialized on CPU while others were already on `cuda:0`. The failure happened before NCCL, at the producer-side `torch.cat`, with mixed CPU and CUDA flattened pieces.

I agree with the intent of your suggested change: the pieces need to be on CUDA before `torch.cat` and NCCL broadcast. The current PR does that through `_normalize_packing_tensors_for_broadcast(...)` immediately before `torch.cat` in both producer flush paths.

The reason I used a helper there is to handle both the normal flush and final flush paths consistently, and to preserve an existing CUDA device if one of the pieces is already CUDA. If all pieces are CPU, it falls back to `torch.cuda.current_device()`. Since the tensors are already contiguous flattened `uint8` views at that point, moving those byte views before concatenation preserves the packed broadcast payload.

If you prefer the conversion to be inline at the tensor creation site, I can adjust it while keeping the same device-selection behavior and tests.

**Comment by youngeunkwon0405** [comment]
> Thanks @youngeunkwon0405 . The case I hit was during non-colocated server-generation refit / weight sync. In that path the producer streams state-dict / adapter tensors into packed broadcast, and some pieces were materialized on CPU while others were already on `cuda:0`. The failure happened before NCCL, at the producer-side `torch.cat`, with mixed CPU and CUDA flattened pieces.

I understand that part, I am curious which specific case that will happen (e.g., xxx model (bf16 or fp8) grpo run using automodel or megatron path with xxx, or else). Because this is the first time seeing that CPU-located tensor passed to the refit path. All model params are expected to be located in the GPU before calling refit. Could you please provide more details on that?

> I agree with the intent of your suggested change: the pieces need to be on CUDA before `torch.cat` and NCCL broadcast. The current PR does that through `_normalize_packing_tensors_for_broadcast(...)` immediately before `torch.cat` in both producer flush paths.
> 
> The reason I used a helper there is to handle both the normal flush and final flush paths consistently, and to preserve an existing CUDA device if one of the pieces is already CUDA. If all pieces are CPU, it falls back to `torch.cuda.current_device()`. Since the tensors are already contiguous flattened `uint8` views at that point, moving those byte views before concatenation preserves the packed broadcast payload.
>
> If you prefer the conversion to be inline at the tensor creation site, I can adjust it while keeping the same device-selection behavior and tests.

I think my suggested oneline change is doing the same. The reason I prefer that inline change is that the refit code performance can be sensitive to host code efficiency so making it simple (use less python lines) can be helpful for the performance. 

Also `.to(torch.device("cuda"))` is equivalent to `.to(torch.device("cuda",  torch.cuda.current_device()))` since the default is following the current device.
